File: app/src/main/java/to/holepunch/bare/android/MainActivity.java
package to.holepunch.bare.android;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import to.holepunch.bare.kit.IPC;
import to.holepunch.bare.kit.RPC;
import to.holepunch.bare.kit.Worklet;

public class MainActivity extends Activity {
  static {
    System.loadLibrary("bare_android");
  }

  Worklet worklet;
  IPC ipc;
  RPC rpc;

  @Override
  public void
  onCreate (Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    worklet = new Worklet();

    try {
      worklet.start("/app.bundle", getAssets().open("app.bundle"));
    } catch (Exception e) {
      throw new RuntimeException(e);
    }

    ipc = new IPC(worklet);

    rpc = new RPC(ipc, (req, error) -> {
      if (req.command.equals("ping")) {
        Log.i("bare", req.data("UTF-8"));

        req.reply("Pong from Android", "UTF-8");
      }
    });

    RPC.OutgoingRequest req = rpc.request("ping");

    req.send("Ping from Android", "UTF-8");

    req.reply("UTF-8", (data, error) -> {
      Log.i("bare", data);
    });
  }

  @Override
  public void
  onPause () {
    super.onPause();

    worklet.suspend();
  }

  @Override
  public void
  onResume () {
    super.onResume();

    worklet.resume();
  }

  @Override
  public void
  onDestroy () {
    super.onDestroy();

    try {
      ipc.close();
      ipc = null;
    } catch (Exception e) {
      throw new RuntimeException(e);
    }

    worklet.terminate();
    worklet = null;
  }
}

File: app/src/main/java/to/holepunch/bare/android/MainActivity.kt
package to.holepunch.bare.android

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.os.Bundle
import to.holepunch.bare.android.voip.VoIPManager
import to.holepunch.bare.kit.Worklet

class MainActivity : Activity() {
  private val voipManager: VoIPManager by lazy {
    VoIPManager(this)
  }

  private lateinit var worklet: Worklet

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    if (checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
      requestPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS), 0)
    }

    voipManager.registerPhoneAccount("Bare Android")

    worklet = Worklet(null)

    try {
      worklet.start("/app.bundle", assets.open("app.bundle"), null)
    } catch (e: Exception) {
      throw RuntimeException(e)
    }
  }

  override fun onPause() {
    super.onPause()

    worklet.suspend()
  }

  override fun onResume() {
    super.onResume()

    worklet.resume()
  }

  override fun onDestroy() {
    super.onDestroy()

    worklet.terminate()
  }
}
